lib/fetch-server-test.js

```diff
@@ -0,0 +1,155 @@
+import isFunction from 'lodash/isFunction';
+import intersection from 'lodash/intersection';
+import difference from 'lodash/difference';
+
+import * as sinon from 'sinon';
+
+const VALID_METHODS = [`GET`, `POST`, `PATCH`, `PUT`, `DELETE`];
+
+/**
+ * Utility class to stub fetch requests with sinon. Instantiate with a set of
+ * route handlers mapping URL (sub)strings to canned response values, or functions
+ * which return canned responses/response values. Values will be serialized as JSON
+ * and returned in a Fetch-like Promise object.
+ *
+ * @example
+ * const fetchServer = new FetchServer({
+ *   '/foo.json': 'bar',
+ *   '/baz/quux': (url, params) => !!params.foo ? 'a' : 'z',
+ * }, {debug: true});
+ * fetchServer.start();
+ * const res = await fetch('https://myserver/foo.json');
+ * const val = await res.json(); // 'bar'
+ * 
+ * @example 
+ * You can also have the following handler types where key is any of the HTTP methods
+ * const handler = {
+ *   '/foo/bar`: {
+ *      GET: {
+ *        'content-type': 'application/json',
+ *        'status': 200,
+ *        'result': {}
+ *      },
+ *      // Takes a non-async function, and creates Promise
+ *      POST: (url, params) => ({
+ *        'content-type': 'text/html',
+ *        'status': 200,
+ *        'result': {}
+ *      })
+ *   }
+ * }
+ */
+export class FetchServer {
+  constructor(handlers, { debug = false } = {}) {
+    this.handlers = handlers;
+    this.debug = debug;
+    this.handle = this.handle.bind(this);
+  }
+
+  handle(url, params) {
+    // default method GET
+    const HTTPmethod = params.method || `GET`;
+    let response;
+    for (const [route, handler] of Object.entries(this.handlers)) {
+
+      // Block 1 : parses body
+      if (url.includes(route)) {
+        if (params && params.body) {
+          try {
+            params = JSON.parse(params.body);
+          } catch (e) {
+            if (e instanceof SyntaxError) {
+              // not JSON parse-able, must be a string
```

Comment:
The "try to parse, handle gracefully if not parseable" approach is something I can see the benefit of for ease of use, but I wonder if we'd get a more predictable test experience if the behavior was more explicitly controlled. What I'm thinking is parse in the default case, where `content-type` is the default `application/json`. But if `content-type` is passed as a different value, say `text/html`, don't parse. This gives the calling test control over whether to parse or not, and you don't have cases where a given handler may return an object or error string and the test has to check that type. I can also think of cases we might not want to be parsed ever but would be parsable, say an arb query string or something that could look like `[]`. Wdyt?

Reply:
Yes, that makes sense. Passing `Content-Type` would work the best. Not sure if we set the `Content-Type` in our APIClient though. But yea we can handle this case.